codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +314,498 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+            }
+        }
+
+        /**
+         * Returns the flow-control size for DATA frames, and 0 for all other frames.
+         */
+        private static final class FlowControlledFrameSizeEstimator implements MessageSizeEstimator {
+
+            static final FlowControlledFrameSizeEstimator INSTANCE = new FlowControlledFrameSizeEstimator();
+
+            private static final class EstimatorHandle implements MessageSizeEstimator.Handle {
+
+                static final EstimatorHandle INSTANCE = new EstimatorHandle();
+
+                @Override
+                public int size(Object msg) {
+                    return msg instanceof Http2DataFrame ? ((Http2DataFrame) msg).flowControlledBytes() : 0;
+                }
+            }
+
+            @Override
+            public Handle newHandle() {
+                return EstimatorHandle.INSTANCE;
+            }
+        }
+
+        /**
+         * {@link ChannelConfig} so that the high and low writebuffer watermarks can reflect the outbound flow control
+         * window, without having to create a new {@link WriteBufferWaterMark} object whenever the flow control window
+         * changes.
+         */
+        private final class Http2StreamChannelConfig extends DefaultChannelConfig {
+
+            // TODO(buchgr): Overwrite the RecvByteBufAllocator. We only need it to implement max messages per read.
+            Http2StreamChannelConfig(Channel channel) {
+                super(channel);
+            }
+
+            /**
+             * @deprecated  Use {@link #getWriteBufferWaterMark()} instead.
+             */
+            @Override
+            @Deprecated
+            public int getWriteBufferHighWaterMark() {
+                int window = (int) min(Integer.MAX_VALUE, outboundFlowControlWindow);
+                return max(0, window);
+            }
+
+            /**
+             * @deprecated  Use {@link #getWriteBufferWaterMark()} instead.
+             */
+            @Override
+            @Deprecated
+            public int getWriteBufferLowWaterMark() {
+                return getWriteBufferHighWaterMark();
+            }
+
+            @Override
+            public MessageSizeEstimator getMessageSizeEstimator() {
+                return FlowControlledFrameSizeEstimator.INSTANCE;
+            }
+
+            @Override
+            public WriteBufferWaterMark getWriteBufferWaterMark() {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public ChannelConfig setMessageSizeEstimator(MessageSizeEstimator estimator) {
+                throw new UnsupportedOperationException();
+            }
+
+            /**
+             * @deprecated  Use {@link #setWriteBufferWaterMark(WriteBufferWaterMark)} instead.
+             */
+            @Override
+            @Deprecated
+            public ChannelConfig setWriteBufferHighWaterMark(int writeBufferHighWaterMark) {
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)